### DISCUSSION TITLE
Soft delay extension

### DISCUSSION THREAD

**Discussion by m-a-x-i-m-z** [pulseq/pulseq#75] [Ideas]
As discussed couple of days [here](https://github.com/pulseq/pulseq/discussions/68#discussioncomment-11413537) we've just implemented the extension named _soft delay_, which allows the interpreter to modify durations of the pure delay blocks (e.g. Pulseq blocks that contain no gradient, RF or ADC events). An additional limitation is that such blocks need to have duration larger than 0, which is needed for the efficient implementation of the gradient checks. 

We expect this feature to be useful to allow the operators to modify TE or TR of the sequences, or other delays e.g in the spectroscopic sequences or trigger delays in case of the cardiac triggering.

The soft delays are identified by the numeric ID and the string 'hint'. The latter is expected to help the interpreter to display the delay parameter on the user interface. Additionally, it is possible to specify an offset that will be added to the number provided by the user, which may be useful to directly specifying TE, TI or TR. Currently the proposed format looks like that: 

```
# Extension specification for soft delays:
# id num offset hint
extension DELAYS 2
1 0 0 TR_delay
```

What is little bit unclear to me, whether we should enforce that the Pulseq programmer also provides the reasonable stepping for the delay (e.g. 0.1 ms for TE or 1 ms for TR, etc). We will probably skip this in the current format revision. 

Please check out the dev branch and let us know what you think. In a mean time we will work on the interpreter to support all this...

**Comment by m-a-x-i-m-z** [comment]
Oh, few more things: we intend to implement the logic that the delay provided by the user will be ADDED to the current block duration. 

Another point is that we may implement, in addition to the `offset` field also the `divisor` field to allow for things like specifying TE in the spin echo sequence or TR in multi-slice sequences.

**Comment by zhengliuX** [comment]
One thing is a little unclear to me. There must be different number is dynamic, which means "soft delays" added to the sequence in different scenarios are also different. 

To display the value, we need to pre-define or add the parameters to the protocol (at least in UIH), then the interpreter will get the modified value from the UI(or actually from the protocol). Due to the unpredictability of the "soft delay" number, it seems hard to implement?

Any thoughts for this?

**Comment by m-a-x-i-m-z** [comment]
My vision with the soft delays was that the values that are set by the user need to be stored in the protocol. On Siemens there are several free parameters. Additionally, there is an array of delay times TD. As for now I was unsure whether to map the delays to TD[0], TD[1] etc or to the free parameters. For the user interface iterations, it would be reasonable to display the text hint that is a part of the soft delay definition to the user, but I'm still little bit wondering how to implement that.

**Reply by zhengliuX** [reply]
OK, I tried some on how to make soft delay editable on the UI. In UI, I mainly have two strategies.

1. In UI, we have a widget called NumberVector, which mean a widget could have multiple numbers, like a vector. I can dynamic modify the number of elements, based on the soft delay number stored in Pulseq files, then display the value on the UI, and get the value from the UI before executing the sequence. One thing is, the label of this widget is uncontrolled, it will have a index as suffix by force, which is uneditable in the sequence. Like TE1, TE2, TE3, if I set the label of this widget as "TE" in the protocol. I can set the text hine as the tooltip, so if users want to know which soft delay it is, there is still a way, but not very straightforward.

2.  The second method is to use two widgets, one is a dropdown box storing all the text hint of soft delay, and the other is a number editor, to dispaly&modify the value based on the selected item in the dropdown box. This is more straightforward, but I have to use two widgets.

Best,
Zheng
